compiler/rustc_codegen_cranelift/src/unsize.rs

```diff
@@ -25,39 +25,70 @@ pub(crate) fn unsized_info<'tcx>(
             .bcx
             .ins()
             .iconst(fx.pointer_type, len.eval_usize(fx.tcx, ParamEnv::reveal_all()) as i64),
-        (&ty::Dynamic(..), &ty::Dynamic(..)) => {
-            // For now, upcasts are limited to changes in marker
-            // traits, and hence never actually require an actual
-            // change to the vtable.
-            old_info.expect("unsized_info: missing old info for trait upcast")
+        (&ty::Dynamic(ref data_a, ..), &ty::Dynamic(ref data_b, ..)) => {
+            let old_info =
+                old_info.expect("unsized_info: missing old info for trait upcasting coercion");
+            if data_a.principal_def_id() == data_b.principal_def_id() {
+                return old_info;
+            }
+            // trait upcasting coercion
+            let principal_a = data_a
+                .principal()
+                .expect("unsized_info: missing principal trait for trait upcasting coercion");
+            let principal_b = data_b
+                .principal()
+                .expect("unsized_info: missing principal trait for trait upcasting coercion");
```

Comment:
Isn't the principal non-existent when the trait object only consists of traits like `Send`, `Sync` or `Sized`?

Reply:
Yes, if the principal is non-existent, either it's not upcasting coercion, like `&dyn Send+Sync` to `&dyn Send` will be a simple case (and will take the early return code path), or it will be rejected by type-checking, like `dyn Foo + Send` to `dyn Send` will be rejected.

I added some comments within the code to clarify.